compiler-rt/lib/msan/msan.h

```diff
@@ -106,7 +107,8 @@ const MappingDesc kMemoryLayout[] = {
     {0x510000000000ULL, 0x600000000000ULL, MappingDesc::APP, "app-2"},
     {0x600000000000ULL, 0x610000000000ULL, MappingDesc::ORIGIN, "origin-1"},
     {0x610000000000ULL, 0x700000000000ULL, MappingDesc::INVALID, "invalid"},
-    {0x700000000000ULL, 0x800000000000ULL, MappingDesc::APP, "app-3"}};
+    {0x700000000000ULL, 0x740000000000ULL, MappingDesc::ALLOCATOR, "allocator"},
```

Comment:
why do we need another enum value?

Reply:
InitShadow is called before the allocator is initialized, so we need a way to tell InitShadow that we want to check that the allocator memory range is available (but not map it, because the allocator will do that).

The existing enum values don't suffice:
- MappingDesc::APP does not call CheckMemoryRangeAvailability()
- MappingDesc::INVALID does call CheckMemoryRangeAvailability(), but it will also be protected, which we don't want
- MappingDesc::SHADOW also calls CheckMemoryRangeAvailability(), but it will map the memory, which we also don't want
- MappingDesc::ALLOCATOR will call CheckMemoryRangeAvailability() but will not map or protect the memory

Reply:
Currently, the kernel might randomly map the app into the same region that the allocator expects (this isn't a problem with low-entropy ASLR, because it never maps into the bottom part of the app region, which is where the allocator typically lives). We therefore need to enforce that the APP and ALLOCATOR regions are disjoint.

Reply:
Would it possible to introduce MappingDesc::ALLOCATOR in a separate patch?

Reply:
Do you mean introduce MappingDesc::ALLOCATOR as an NFC patch?

Reply:
yes, and related conditions

Reply:
Ok, will do.

Reply:
Created https://github.com/llvm/llvm-project/pull/85153